topdown/eval.go

```diff
@@ -1696,7 +1742,14 @@ func (e evalBuiltin) eval(iter unifyIterator) error {
 				err = iter()
 			}
 		} else {
-			err = e.e.unify(e.terms[len(e.terms)-1], output, iter)
+			err = e.e.unify(e.terms[endIndex], output, iter)
+		}
+
+		// If the NDBCache is present, we can assume this builtin
+		// call was not cached earlier.
+		if e.canUseNDBCache(e.bi) {
+			// Populate the NDBCache from the output term.
+			e.bctx.NDBuiltinCache.Put(e.bi.Name, ast.NewArray(e.terms[:endIndex]...), output.Value)
 		}
```

Comment:
I moved this block out of the 3rd unification case to address the issues brought up by @srenatus with his excellent test case that included a rule body that discarded the builtin result. This should be a safe change because we will never hit this block unless the cache lookup earlier got no results.